### PR TITLE
[PoC] Multiple Callbacks for Base Functionality

### PR DESCRIPTION
**Pull request by soumik12345**
## Callbacks

- [x] Metrics Logger
- [x] Model Checkpoint
- [x] Gradient Accumulation
- [x] Model Graph Logging
- [x] FLOPs

### CODE REVIEW

**Comment by ayulockin** [comment]
What's the idea of breaking down `BaseWandbCallback` into smaller callbacks? 

How should the end user use them? - `from wandb.keras.callbacks import WandbMetricLogger`?

Will they be used internally by the `BaseWandbCallback`?

(PS: I do understand that it is more inline with Keras way of doing things? Asking from a user pod.)

cc: @soumik12345

**Comment by soumik12345** [comment]
> What's the idea of breaking down `BaseWandbCallback` into smaller callbacks?
> 
> How should the end user use them? - `from wandb.keras.callbacks import WandbMetricLogger`?
> 
> Will they be used internally by the `BaseWandbCallback`?
> 
> (PS: I do understand that it is more inline with Keras way of doing things? Asking from a user pod.)
> 
> cc: @soumik12345

@ayulockin

The idea is to follow the keras design philosophy of letting the user decide exactly which feature to use. Granted that the same level of control can be achieved using `wandb.keras.WandbCallback`, however, I think breaking it down into multiple callbacks would not only make it easier and less convoluted for the user to use but also encourage users to take advantage of specific features and build upon them when necessary.

Here's a sample usage:

```python
import tensorflow as tf

import wandb
from wandb.keras.callbacks import (
    WandBMetricsLogger,
    WandbModelCheckpointCallback,
    WandbGradientLogger,
    ModelLogger,
    FLOPsLogger,
)


with wandb.init(project="mnist", job_type="dev-wandb-metrics-logger"):

    fashion_mnist = tf.keras.datasets.fashion_mnist

    (train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()
    train_images = train_images / 255.0
    test_images = test_images / 255.0

    model = tf.keras.Sequential(
        [
            tf.keras.layers.Flatten(input_shape=(28, 28)),
            tf.keras.layers.Dense(128, activation="relu"),
            tf.keras.layers.Dense(10),
        ]
    )

    model.compile(
        optimizer="adam",
        loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
        metrics=["accuracy"],
    )

    callbacks = [
        WandBMetricsLogger(),
        WandbModelCheckpointCallback(save_model_as_artifact=True),
        WandbGradientLogger(
            training_data=(train_images, train_labels), input_shape=(28, 28)
        ),
        ModelLogger(),
        FLOPsLogger(log_flops_on_workspace=True),
    ]

    model.fit(
        train_images,
        train_labels,
        epochs=10,
        callbacks=callbacks,
    )
```

**Comment by ayulockin** [comment]
> Granted that the same level of control can be achieved using wandb.keras.WandbCallback, I think breaking it down into multiple callbacks would not only make it easier and less convoluted

@soumik12345, this drives home the point.

File: wandb/integration/keras/__init__.py [hunk 1]
```diff
@@ -5,5 +5,16 @@
 """
 
 from .keras import WandbCallback
+from .callbacks import WandBMetricsLogger
+from .callbacks import WandbModelCheckpointCallback
+from .callbacks import WandbGradientLogger
+from .callbacks import ModelLogger
```
Comment (ayulockin) [inline wandb/integration/keras/__init__.py]: Can't we also call it `WandbModelLogger` for naming consistency?

File: wandb/integration/keras/callbacks/__init__.py [hunk 2]
```diff
@@ -0,0 +1,4 @@
+from .metrics_logger import WandBMetricsLogger
+from .model_checkpoint import WandbModelCheckpointCallback
+from .gradient_logger import WandbGradientLogger
+from .model_logger import ModelLogger
```
Comment (ayulockin) [inline wandb/integration/keras/callbacks/__init__.py]: same `WandbModelLogger`

File: wandb/keras/__init__.py [hunk 3]
```diff
@@ -6,5 +6,15 @@
 """
 
 from wandb.integration.keras import WandbCallback  # type: ignore
+from wandb.integration.keras import WandBMetricsLogger  # type: ignore
+from wandb.integration.keras import WandbModelCheckpointCallback  # type: ignore
+from wandb.integration.keras import WandbGradientLogger  # type: ignore
+from wandb.integration.keras import ModelLogger  # type: ignore
```
Comment (ayulockin) [inline wandb/keras/__init__.py]: same `WandbModelLogger`

File: wandb/integration/keras/callbacks/metrics_logger.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+    def on_epoch_end(self, epoch, logs={}):
+        wandb.log({"epoch": epoch}, commit=False)
+        wandb.log(logs, commit=True)
+
+    def on_batch_end(self, batch, logs={}):
+        if self.log_batch_frequency and batch % self.log_batch_frequency == 0:
+            wandb.log(logs, commit=True)
+
+    def on_train_batch_end(self, batch, logs={}):
```
Comment (ayulockin) [inline wandb/integration/keras/callbacks/metrics_logger.py:25]: Do we have `on_batch_end` for backward compatibility?

File: wandb/integration/keras/callbacks/metrics_logger.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+from tensorflow.keras import callbacks
+
+from ..keras import patch_tf_keras
+
+
+patch_tf_keras()
+
+
+class WandBMetricsLogger(callbacks.Callback):
```
Comment (ayulockin) [inline wandb/integration/keras/callbacks/metrics_logger.py:12]: - [ ] Can we name it `WandbMetricsLogger` (not use captial B) to keep it consistent with `WandbCallback`?
- [ ] I compared this callback with `WandbCallback`. I noticed that the system metrics are not updating properly.
    * This [run page](https://wandb.ai/ayut/mirbind/runs/aeotj3bl/system?workspace=user-ayut) uses `WandbCallback`.
    * This [run page](https://wandb.ai/ayut/mirbind/runs/gx0bzbwa/system?workspace=) uses `WandBMetricsLogger`. Notice that the system metrics have not updated properly.

File: wandb/integration/keras/callbacks/model_logger.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+import tensorflow as tf
+from tensorflow.keras import callbacks
+
+from ..keras import patch_tf_keras
+
+patch_tf_keras()
+
+
+class ModelLogger(callbacks.Callback):
```
Comment (ayulockin) [inline wandb/integration/keras/callbacks/model_logger.py]: I think we should move this to `WandBMetricsLogger`. I don't think as a user I would use this separately cz `model.summary` is itself a great feature.

File: wandb/integration/keras/callbacks/model_checkpoint.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+from tensorflow.keras import callbacks
+
+from ..keras import patch_tf_keras
+
+
+patch_tf_keras()
+
+
+class WandbModelCheckpointCallback(callbacks.Callback):
```
Comment (ayulockin) [inline wandb/integration/keras/callbacks/model_checkpoint.py:13]: * I ran this callback with default setting. I noticed that the model is logged as W&B file in the `.h5` format. [run page](https://wandb.ai/ayut/mirbind/runs/rb70gkyn/files?workspace=user-ayut)
* When I used `save_weights_only=True` there wasn't anything logged even as W&B file. [run page](https://wandb.ai/ayut/mirbind/runs/3vsl41ti/files?workspace=)
* Using `save_model_as_artifact=True` worked as expected.
* Using both `save_weights_only=True` and `save_model_as_artifact=True`  gave this error: `ValueError: Path is not a directory: /home/ayusht/wandb-cli/client/wandb/run-20220817_080253-35ry19kd/files/model-best`.

File: wandb/integration/keras/callbacks/model_checkpoint.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        name = wandb.util.make_artifact_name_safe(f"model-{wandb.run.name}")
+        model_artifact = wandb.Artifact(name, type="model")
+        model_artifact.add_dir(self.filepath[:-3])
+        wandb.run.log_artifact(model_artifact, aliases=["latest", f"epoch_{epoch}"])
+
+        # Remove the SavedModel from wandb dir as we don't want to log it to save memory.
+        shutil.rmtree(self.filepath[:-3])
+
+    def on_epoch_end(self, epoch, logs=None):
+        self._save_model(epoch)
```
Comment (ayulockin) [inline wandb/integration/keras/callbacks/model_checkpoint.py:75]: I think we should remove this and just stick with saving the model as artifact.

File: wandb/integration/keras/callbacks/model_checkpoint.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+patch_tf_keras()
+
+
+class WandbModelCheckpointCallback(callbacks.Callback):
+    def __init__(
+        self,
+        save_weights_only: bool = False,
+        save_model_as_artifact: bool = False,
+        verbose: int = 0,
+        *args,
```
Comment (ayulockin) [inline wandb/integration/keras/callbacks/model_checkpoint.py:19]: Can we have an argument to set checkpoint frequency?

File: wandb/integration/keras/callbacks/model_checkpoint.py (hunk 7 above)
Comment (ayulockin) [inline wandb/integration/keras/callbacks/model_checkpoint.py:13]: I was thinking we should try to subclass `tf.keras.callbacks.ModelCheckpoint` to take care of the checkpointing logic and use `wandb.Artifact` to log the model.

File: wandb/integration/keras/callbacks/model_logger.py [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._graph_rendered = False
+
+    def _render_graph(self):
+        if not self._graph_rendered:
+            wandb.run.summary["graph"] = wandb.Graph.from_keras(self.model)
```
Comment (ayulockin) [inline wandb/integration/keras/callbacks/model_logger.py:18]: I think we should have a `try-except` clause here. Check out this jira [ticket ](https://wandb.atlassian.net/browse/WB-10266?focusedCommentId=46319&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-46319) for the edge case.

File: wandb/integration/keras/callbacks/gradient_logger.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+from ..keras import patch_tf_keras, _GradAccumulatorCallback, _CustomOptimizer
+
+
+tf_logger = tf.get_logger()
+
+patch_tf_keras()
+
+
+class WandbGradientLogger(tf.keras.callbacks.Callback):
```
Comment (ayulockin) [inline wandb/integration/keras/callbacks/gradient_logger.py:16]: I was not able to use it, 

trying to do this: `WandbGradientLogger(training_data=x_train[:1000], input_shape=(None, 784))`

File: wandb/integration/keras/callbacks/gradient_logger.py [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    def __init__(
+        self,
+        training_data,
+        input_shape: Optional[Sequence[int]] = None,
```
Comment (ayulockin) [inline wandb/integration/keras/callbacks/gradient_logger.py:20]: Can't we get `input_shape` from the model instance?

File: wandb/integration/keras/callbacks/gradient_logger.py (hunk 11 above)
Comment (ayulockin) [reply wandb/integration/keras/callbacks/gradient_logger.py:16]: Also not sure if Keras users will use it separately or should it be part of the metric logger class. Maybe we have have this class but will be invoked from the `WandbMetricsLogger`. But for now we can keep it as it is.

**Review by ayulockin — CHANGES_REQUESTED** [review]

**Comment by ayulockin** [comment]
I see that FLOPs is ticked, but I can't find it in the changes. Did you miss to commit the callback? :P cc: @soumik12345

File: wandb/integration/keras/callbacks/metrics_logger.py (hunk 5 above)
Comment (soumik12345) [reply wandb/integration/keras/callbacks/metrics_logger.py:12]: Hi @ayulockin, here's the [run page](https://wandb.ai/geekyrakshit/mnist/runs/p9pryo70/system) of my testing the `WandBMetricsLogger`. Here's the code I used, seems to be working fine:

```python
import tensorflow as tf

import wandb
from wandb.keras import (
    WandBMetricsLogger,
    WandbModelCheckpointCallback,
    WandbGradientLogger,
    WandbModelLogger,
    FLOPsLogger,
)


with wandb.init(project="mnist", job_type="dev-wandb-metrics-logger"):

    fashion_mnist = tf.keras.datasets.fashion_mnist

    (train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()
    train_images = train_images / 255.0
    test_images = test_images / 255.0

    class_names = [
        "T-shirt/top",
        "Trouser",
        "Pullover",
        "Dress",
        "Coat",
        "Sandal",
        "Shirt",
        "Sneaker",
        "Bag",
        "Ankle boot",
    ]

    model = tf.keras.Sequential(
        [
            tf.keras.layers.Flatten(input_shape=(28, 28)),
            tf.keras.layers.Dense(128, activation="relu"),
            tf.keras.layers.Dense(10),
        ]
    )

    model.compile(
        optimizer="adam",
        loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
        metrics=["accuracy"],
    )

    callbacks = [
        WandBMetricsLogger(),
    ]

    model.fit(
        train_images,
        train_labels,
        epochs=10,
        callbacks=callbacks,
    )
```

File: wandb/integration/keras/callbacks/metrics_logger.py (hunk 4 above)
Comment (soumik12345) [reply wandb/integration/keras/callbacks/metrics_logger.py:25]: Yes.

File: wandb/integration/keras/callbacks/model_logger.py (hunk 6 above)
Comment (soumik12345) [reply wandb/integration/keras/callbacks/model_logger.py]: Done!!!

File: wandb/integration/keras/callbacks/model_checkpoint.py (hunk 7 above)
Comment (soumik12345) [reply wandb/integration/keras/callbacks/model_checkpoint.py:13]: On it...

Comment (soumik12345) [reply wandb/integration/keras/callbacks/model_checkpoint.py:13]: Great idea!!!

**Comment by ayulockin** [comment]
Hey @soumik12345 I have created a minimal checkpointing callback by subclassing `tf.keras.callbacks.ModelCheckpoint` and is baked with W&B Artifacts.

```
class WandbModelCheckpoint(tf.keras.callbacks.ModelCheckpoint):
    def __init__(self,
                 filepath='wandb/model_{epoch}',
                 monitor='val_loss',
                 save_best_only=False,
                 save_weights_only=False,
                 initial_value_threshold=None,
                 **kwargs):
        super(WandbModelCheckpoint, self).__init__(
            filepath,
            monitor,
            save_best_only,
            save_weights_only,
            initial_value_threshold,
            **kwargs
        )

    def on_epoch_end(self, epoch, logs=None):
        super().on_epoch_end(epoch, logs)
        # Get filepath where the model checkpoint is saved.
        filepath = self._get_file_path(epoch, batch=None, logs=logs)
        # Log the model as artifact
        self._log_ckpt_as_artifact(filepath)

    def _log_ckpt_as_artifact(self, filepath):
        """Log model checkpoint as  W&B Artifact."""
        model_artifact = wandb.Artifact(
            f'run_{wandb.run.id}_model', type='model')
        model_artifact.add_dir(filepath)
        wandb.log_artifact(model_artifact)
```

Hope it helps.

**Pull request merged**